File: src/Solutions/Euler112.hs
module Solutions.Euler112 where
import Utils.Operators((<||>))
import Control.Monad
import Control.Monad.ST
import Data.STRef

inc :: String -> Bool
inc [c] = True
inc (h1:h2:t) = h1 <= h2 && inc (h2:t)
dec :: String -> Bool
dec = inc . reverse
isBouncy :: Int -> Bool
isBouncy = not . (inc <||> dec) . show

compute = runST $ do
    n <- newSTRef 99
    bouncy <- newSTRef 0
    let loop = do
            n' <- readSTRef n
            bouncy' <- readSTRef bouncy
            when (100 * bouncy' < 99 * n') $ do
                when (isBouncy $ n' + 1) $ modifySTRef bouncy succ
                modifySTRef n succ
                loop
    loop
    readSTRef n

euler112 :: IO String
euler112 = return $ show compute

File: rust/src/solutions/euler112.rs
use crate::utils::numeric::digits::Digits;

fn is_bouncy(n: usize) -> bool {
    let digits: Vec<usize> = n.digits();
    let mut not_inc = false;
    let mut not_dec = false;
    for i in 1..digits.len() {
        not_inc = not_inc || digits[i] > digits[i - 1];
        not_dec = not_dec || digits[i] < digits[i - 1];
    }
    not_inc && not_dec
}

pub fn euler112() -> String {
    let mut n = 99;
    let mut bouncy = 0;
    while 100 * bouncy != 99 * n {
        if is_bouncy(n + 1) {
            bouncy += 1;
        }
        n += 1;
    }
    n.to_string()
}
